hardware_interface/src/resource_manager.cpp

```diff
@@ -51,116 +81,416 @@ class ResourceStorage
+            " This can happen due to multiple calls to 'cleanup'",
+            hardware.get_name().c_str(), interface.c_str());
+        }
+      }
+      // remove all state interfaces from available list
+      for (const auto & interface : hardware_info_map_[hardware.get_name()].state_interfaces)
+      {
+        auto found_it = std::find(
+          available_state_interfaces_.begin(), available_state_interfaces_.end(), interface);
+
+        if (found_it != available_state_interfaces_.end())
+        {
+          available_state_interfaces_.erase(found_it);
+          // TODO(destogl): change output to debug
+          RCUTILS_LOG_INFO_NAMED(
+            "resource_manager", "(hardware '%s'): '%s' state interface removed from available list",
+            hardware.get_name().c_str(), interface.c_str());
+        }
+        else
+        {
+          // TODO(destogl): do here error management if interfaces are only partially added into
+          // "available" list - this should never be the case!
+          RCUTILS_LOG_WARN_NAMED(
+            "resource_manager",
+            "(hardware '%s'): '%s' state interface not in available list. "
+            "This can happen due to multiple calls to 'cleanup'",
+            hardware.get_name().c_str(), interface.c_str());
+        }
+      }
+    }
+    return result;
+  }
+
+  template <class HardwareT>
+  bool shutdown_hardware(HardwareT & hardware)
+  {
+    bool result = trigger_hardware_state_transition(
+      std::bind(&HardwareT::shutdown, &hardware), "shutdown", hardware.get_name(),
+      lifecycle_msgs::msg::State::PRIMARY_STATE_UNCONFIGURED);
+
+    if (result)
+    {
+      // TODO(destogl): change this - deimport all things if there is there are interfaces there
+      // deimport_non_movement_command_interfaces(hardware);
+      // deimport_state_interfaces(hardware);
+    }
+    return result;
+  }
+
+  template <class HardwareT>
+  bool activate_hardware(HardwareT & hardware)
+  {
+    bool result = trigger_hardware_state_transition(
+      std::bind(&HardwareT::activate, &hardware), "activate", hardware.get_name(),
+      lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE);
+
+    if (result)
+    {
+      // TODO(destogl): make all command interfaces available (currently are
+      // all available)
+    }
+
+    return result;
+  }
+
+  template <class HardwareT>
+  bool deactivate_hardware(HardwareT & hardware)
+  {
+    bool result = trigger_hardware_state_transition(
+      std::bind(&HardwareT::deactivate, &hardware), "deactivate", hardware.get_name(),
+      lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE);
+
+    if (result)
+    {
+      // TODO(destogl): make all command interfaces unavailable that should be present only
+      // when active (currently are all available) also at inactive
+    }
+    return result;
+  }
+
+  template <class HardwareT>
+  bool set_component_state(HardwareT & hardware, const rclcpp_lifecycle::State & target_state)
```

Comment:
can't we inherit from lifecycle node? plus I really think that this should not live in resource_manager but somewhere in the components base files

Reply:
Agree, but then we are going back to adding additional layer for different hardware types, or we should repeat this three times.

We are inheriting from Lifecycle Node interface and there is no such implementation there.

On the other hand this is a convenience method, and it is simply running lifecycle if we have to “jump” two states to achieve goal state. I don't see necessary this misplaced here because ResourceManager has to be very-well aware of the hardware lifecycle.

Reply:
yeah but I was just wondering if something like this could go to a separate header file as it does very specific stuff

Reply:
https://github.com/ros-controls/ros2_control/issues/626